File: plugins/rider/src/main/kotlin/com/shortsleevestudio/gamescript/handlers/NotificationHandlers.kt
package com.shortsleevestudio.gamescript.handlers

import com.google.gson.JsonObject
import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.wm.WindowManager
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

private const val DEFAULT_STATUS_TIMEOUT_MS = 3000

/**
 * Handlers for notification operations.
 * Implements: notify, status
 *
 * Uses non-modal balloon notifications (like VSCode's toasts) rather than
 * blocking modal dialogs for better UX consistency.
 */
class NotificationHandlers(private val context: HandlerContext) {

    // Scheduled executor for status bar timeout - avoids blocking pooled threads
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    /**
     * Get all handlers as a map.
     */
    fun getHandlers(): Map<String, suspend (JsonObject) -> Unit> = mapOf(
        "notify" to ::handleNotify,
        "status" to ::handleStatus
    )

    /**
     * Handle notify message - shows balloon notification.
     * Uses IntelliJ's Notifications API for non-blocking toasts (like VSCode).
     */
    private suspend fun handleNotify(message: JsonObject) {
        val text = message.get("message")?.asString ?: return
        val level = message.get("level")?.asString ?: "info"
        val detail = message.get("detail")?.asString

        val fullMessage = if (detail != null) "$text\n\n$detail" else text

        val notificationType = when (level.lowercase()) {
            "error" -> NotificationType.ERROR
            "warning" -> NotificationType.WARNING
            else -> NotificationType.INFORMATION
        }

        ApplicationManager.getApplication().invokeLater {
            NotificationGroupManager.getInstance()
                .getNotificationGroup("GameScript")
                .createNotification(fullMessage, notificationType)
                .notify(context.project)
        }
    }

    /**
     * Handle status message - shows text in status bar temporarily.
     * Uses timeoutMs parameter name to match shared types.
     */
    private suspend fun handleStatus(message: JsonObject) {
        val text = message.get("message")?.asString ?: return
        // Support both 'timeoutMs' (from shared types) and 'timeout' (legacy)
        val timeout = message.get("timeoutMs")?.asInt
            ?: message.get("timeout")?.asInt
            ?: DEFAULT_STATUS_TIMEOUT_MS

        ApplicationManager.getApplication().invokeLater {
            val statusBar = WindowManager.getInstance().getStatusBar(context.project)
            statusBar?.info = text

            // Clear after timeout using scheduled executor (non-blocking)
            if (timeout > 0) {
                scheduler.schedule({
                    ApplicationManager.getApplication().invokeLater {
                        if (statusBar?.info == text) {
                            statusBar.info = ""
                        }
                    }
                }, timeout.toLong(), TimeUnit.MILLISECONDS)
            }
        }
    }

    /**
     * Cleanup resources when handler is no longer needed.
     */
    fun dispose() {
        scheduler.shutdown()
    }
}

File: plugins/dotnet/visual-studio/GameScript.VisualStudio/Handlers/NotificationHandlers.cs
using System.Text.Json;
using Microsoft.VisualStudio.Extensibility.Shell;

namespace GameScript.VisualStudio;

/// <summary>
/// Handlers for notification operations.
/// Implements: notify, status
///
/// Uses non-modal prompts and status bar messages for better UX consistency.
/// </summary>
public static class NotificationHandlers
{
    private const int DefaultStatusTimeoutMs = 3000;

    public static void Register(MessageMediator mediator, HandlerContext context)
    {
        mediator.Register("notify", (msg, ct) => HandleNotify(msg, context, ct));
        mediator.Register("status", (msg, ct) => HandleStatus(msg, context, ct));
    }

    /// <summary>
    /// Handle notify message - shows prompt notification.
    /// Uses Visual Studio's ShowPromptAsync for user notifications.
    /// </summary>
    private static async Task HandleNotify(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        var text = message.GetOptionalString("message");
        if (string.IsNullOrEmpty(text))
        {
            return;
        }

        var level = message.GetOptionalString("level") ?? "info";
        var detail = message.GetOptionalString("detail");

        var fullMessage = !string.IsNullOrEmpty(detail) ? $"{text}\n\n{detail}" : text;

        try
        {
            // ShowPromptAsync displays a message to the user
            // For errors/warnings, we use the prompt; for info, we could use status bar
            // but prompts ensure the user sees important messages
            if (level.Equals("error", StringComparison.OrdinalIgnoreCase) ||
                level.Equals("warning", StringComparison.OrdinalIgnoreCase))
            {
                await context.Extensibility.Shell().ShowPromptAsync(fullMessage, PromptOptions.OK, ct);
            }
            else
            {
                // For info/success, use status bar which is less intrusive
                await ShowStatusWithTimeoutAsync(context, fullMessage, DefaultStatusTimeoutMs, ct);
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[NotificationHandlers] Error showing notification: {ex.Message}");
        }
    }

    /// <summary>
    /// Handle status message - shows text in status bar temporarily.
    /// Uses timeoutMs parameter name to match shared types.
    /// </summary>
    private static async Task HandleStatus(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        var text = message.GetOptionalString("message");
        if (string.IsNullOrEmpty(text))
        {
            return;
        }

        // Support both 'timeoutMs' (from shared types) and 'timeout' (legacy)
        var timeout = message.GetOptionalInt("timeoutMs")
            ?? message.GetOptionalInt("timeout")
            ?? DefaultStatusTimeoutMs;

        try
        {
            await ShowStatusWithTimeoutAsync(context, text, timeout, ct);
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[NotificationHandlers] Error showing status: {ex.Message}");
        }
    }

    /// <summary>
    /// Show a status bar message that auto-clears after the specified timeout.
    /// </summary>
    private static async Task ShowStatusWithTimeoutAsync(
        HandlerContext context,
        string message,
        int timeoutMs,
        CancellationToken ct)
    {
        // ShowStatusMessageAsync returns a disposable that clears the message when disposed
        using var status = await context.Extensibility.Shell().ShowStatusMessageAsync(message, ct);

        if (timeoutMs > 0)
        {
            // Wait for the timeout, then the using block will dispose and clear the message
            await Task.Delay(timeoutMs, ct);
        }
    }
}
